### PR TITLE
llama3 rope

### LINKED ISSUE #39: [feat] Llama 3.x rope scaling support
# 🧐 Problem Description

Fast-LLM lacks support for Llama 3.x models due to missing compatibility with Llama-3-style RoPE scaling. This prevents us from effectively training or using Llama 3.x checkpoints on long contexts.

To support Llama 3's full long-context pretraining (up to 128k tokens), Fast-LLM eventually needs to implement RoPE scaling. This includes handling scaling parameters like `factor`, `low_freq_factor`, and `high_freq_factor`, which allow positional embeddings to adapt for long sequences. 

As an **interim solution**, Fast-LLM could ignore the `rope_scaling` dictionary and proceed with training for contexts up to 8k tokens, allowing basic Llama 3 compatibility without the need for immediate scaling support.

Example config for Llama 3.x models:

```json
{
  "architectures": [
    "LlamaForCausalLM"
  ],
  ...
  "rope_scaling": {
    "factor": 8.0,
    "low_freq_factor": 1.0,
    "high_freq_factor": 4.0,
    "original_max_position_embeddings": 8192,
    "rope_type": "llama3"
  },
  ...
}
```

# 💡 Proposed Solution

1. **Interim Solution**: Ignore the `rope_scaling` dictionary in the configuration and proceed with standard 8k token contexts for immediate compatibility with Llama 3.x. This enables Llama 3 models to be trained with the default positional embeddings.

2. **Long-Term Solution**: Implement full Llama-3-style RoPE scaling in Fast-LLM to support training and inference with extended context windows up to 128k tokens. This would involve adding support for the `rope_scaling` parameters and applying these consistently across training and inference.

# 🔄 Alternatives Considered

Not supporting Llama 3's long-context capabilities would limit Fast-LLM’s compatibility with these models. While ignoring the RoPE scaling parameters enables immediate training on up to 8k tokens, adding full RoPE scaling support is necessary for training on longer contexts.

# 📈 Potential Benefits

- **Immediate Compatibility**: The interim solution provides basic compatibility with Llama 3.x, allowing Fast-LLM to train models with up to 8k context lengths.
- **Future-Ready**: Implementing full RoPE scaling will enable Fast-LLM to support long-context adaptation, making it suitable for tasks requiring large context windows.
- **Critical for StarDoc**: This is especially important for projects like StarDoc, which depend on Llama 3.x support and effective handling of extended contexts.

# 📝 Additional Context

For reference, the RoPE scaling mechanism is implemented here:

https://github.com/huggingface/transformers/blob/3ea3ab62d80d91f9bdd16bd3cacd8133fb0d4566/src/transformers/modeling_rope_utils.py#L310-L350

### PR DESCRIPTION
**Pull request by RaymondLi0**
# ✨ Description

Closes #39 

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [x] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

- add llama3-style rope scaling
- add llama3 test config (run with `MODEL=llama3 pytest ./tests/test_checkpoint.py`)
- handle nested HF configs in conversion

### Testing

`MODEL=llama3 pytest ./tests/test_checkpoint.py` passes.
When "sabotaging" the conversion here https://github.com/ServiceNow/Fast-LLM/blob/fc9ef1388fd79b1f74daa5b2ad30deec9c122685/fast_llm/models/gpt/conversion.py#L251-L270
with
```python
def export_rotary_scaling_type(fast_llm_value: RotaryScalingType) -> dict[str, typing.Any] | None:
    return None


def import_rotary_scaling_type(export_value):
    return RotaryScalingType.llama3
```
then `test_run_converted_model` fails, showing that model outputs differ when using an incorrect rotary config.

### CODE REVIEW

**Comment by tscholak** [comment]
Hi @RaymondLi0! Functionally this looks like what we want (pending model conversion), but are you confident (i.e. have you checked) that the forward and backward passes of hf-llama and fast-llm-llama are the same?

File: fast_llm/layers/transformer/config.py [hunk 1]
```diff
@@ -127,6 +127,11 @@ class TransformerArchitectureConfig(BaseModelArchitectureConfig):
         desc="Scale for the rotary positional embeddings. Default: -math.log(10000) = -9.210",
         hint=FieldHint.feature,
     )
+    rotary_scaling_type: RotaryScalingType = Field(
```
Comment (jlamypoirier) [inline fast_llm/layers/transformer/config.py:130]: how about `rotary_embedding_type` so we make it more general?
Or better, merge with `use_rotary_embeddings` into `rotary_embeddings: RotaryEmbeddingType | None`

File: fast_llm/models/gpt/conversion.py [hunk 2]
```diff
@@ -258,6 +258,7 @@ def _create_config_converters(cls) -> list[ParamConverter]:
             # TODO: Llama supports biases
             ConstantExportParamConverter(None, "attention_bias", False),
             ConstantExportParamConverter(None, "mlp_bias", False),
+            # TODO: Convert rope_scaling (llama3 scaling)
```
Comment (jlamypoirier) [inline fast_llm/models/gpt/conversion.py:261]: Are you planning to do in a separate PR?

**Review by jlamypoirier — COMMENTED** [review]
I didn't review for correctness, but looks ok. I'll leave the approval to @tscholak

File: fast_llm/models/gpt/conversion.py (hunk 2 above)
Comment (RaymondLi0) [reply fast_llm/models/gpt/conversion.py:261]: Yes, if that's ok

**Comment by RaymondLi0** [comment]
> Hi @RaymondLi0! Functionally this looks like what we want (pending model conversion), but are you confident (i.e. have you checked) that the forward and backward passes of hf-llama and fast-llm-llama are the same?

Haven't done that check. Do we have existing tests comparing forward/backward of fast-llm and transformers? If no I can look into adding this

**Comment by jlamypoirier** [comment]
> Haven't done that check. Do we have existing tests comparing forward/backward of fast-llm and hf-transformers? If no I can look into adding this

There is one in test_checkpoint https://github.com/ServiceNow/Fast-LLM/blob/main/tests/test_checkpoint.py#L31. It could work for this one if we added a llama3 model to the testing suite (in `common.py`)

File: fast_llm/layers/transformer/config.py (hunk 1 above)
Comment (RaymondLi0) [reply fast_llm/layers/transformer/config.py:130]: Would make sense.
Although that would make this a breaking change right? Would that be OK?

Comment (RaymondLi0) [reply fast_llm/layers/transformer/config.py:130]: @tscholak @jlamypoirier I could merge into `rotary_embeddings: RotaryEmbeddingType | None`, and keep `use_rotary_embeddings` as a deprecated argument to maintain backwards compatibility. wdyt?

Comment (tscholak) [reply fast_llm/layers/transformer/config.py:130]: ok!

Comment (RaymondLi0) [reply fast_llm/layers/transformer/config.py:130]: Hmm, in addition to breaking existing checkpoints it will make conversion more complex than needed. Let's postpone to another PR?

Comment (tscholak) [reply fast_llm/layers/transformer/config.py:130]: Sounds good! We only want to unblock model conversion at this point. Thanks!

**Review by tscholak — APPROVED** [review]
LGTM, thanks @RaymondLi0!

**Comment by tscholak** [comment]
@RaymondLi0 @jlamypoirier merge?

**Comment by jlamypoirier** [comment]
Looks ok, I'll do some adjustments and then merge.

File: tests/common.py [hunk 3]
```diff
@@ -161,6 +176,7 @@
         CONFIG_MIXTRAL_COMMON,
         MixtralGPTHuggingfaceCheckpointFormat,
     ),
+    "llama3": ("gpt", CONFIG_LLAMA3_FAST_LLM, CONFIG_LLAMA3_MEGATRON, CONFIG_LLAMA3_COMMON, LlamaGPTHuggingfaceCheckpointFormat),  # TODO
```
Comment (jlamypoirier) [inline tests/common.py:179]: Why the todo?

**Comment by jlamypoirier** [comment]
Main changes I did:
* Adjust names (simplify, avoid abbreviations)
* Moved nested dict functions to utils since they are useful elsewhere, generalized them a bit.
* Moved rotary preprocessing to `layers` module so we can use the rotary config instead of a ton of arguments
* Replaced rotary scaling type with rotary type, also serves as rotary enabling parameter
* Replaced rotary log scale with theta (wanted to do for a while, now is a good time)
* Moved remaining rotary parameters to new rotary config where they belong
* Added backward compatibility for the modified config parameters

File: tests/common.py (hunk 3 above)
Comment (RaymondLi0) [reply tests/common.py]: Forgot to remove it

**Review by jlamypoirier — APPROVED** [review]
Got the tests to pass for both default and llama3 models. Not sure if that's enough? (Could try loading an old checkpoint to be sure)

**Comment by tscholak** [comment]
I'll continue the training of Llama 3.1 8B with this PR as a further test.

**Comment by tscholak** [comment]
This may need revision.

I am currently cpt-ing Llama 3.1 8B on LambdaLabs, and I see this in `config.yaml` in the output folder:

```yaml
model:
  base_model:
    cross_entropy_impl: fused
    init_method_std_embed: 0.015625
    tie_word_embeddings: false
    transformer:
      activation_type: silu
      add_linear_biases: false
      ffn_hidden_size: 14336
      gated: true
      head_groups: 8
      hidden_size: 4096
      init_method_std: 0.015625
      init_method_std_attn_proj: 0.001953125
      init_method_std_mlp_1: 0.015625
      init_method_std_mlp_2: 0.001953125
      init_method_std_qkv: 0.015625
      kv_channels: 128
      mlp_lr_scale:
      - null
      normalization:
        type: rms_norm
      num_attention_heads: 32
      num_layers: 32
      rotary:
        theta: 500000.0
        type: default
    use_position_embeddings: false
    vocab_size: 128256
```

Instead of `type: default` in the `model.base_model.transformer.rotary`, I expected to see `llama`.

**Review by tscholak — CHANGES_REQUESTED** [review]
This needs revision, because it doesn't work correctly.

I am currently cpt-ing Llama 3.1 8B on LambdaLabs, and I see this in `config.yaml` in the output folder:

```yaml
model:
  base_model:
    cross_entropy_impl: fused
    init_method_std_embed: 0.015625
    tie_word_embeddings: false
    transformer:
      activation_type: silu
      add_linear_biases: false
      ffn_hidden_size: 14336
      gated: true
      head_groups: 8
      hidden_size: 4096
      init_method_std: 0.015625
      init_method_std_attn_proj: 0.001953125
      init_method_std_mlp_1: 0.015625
      init_method_std_mlp_2: 0.001953125
      init_method_std_qkv: 0.015625
      kv_channels: 128
      mlp_lr_scale:
      - null
      normalization:
        type: rms_norm
      num_attention_heads: 32
      num_layers: 32
      rotary:
        theta: 500000.0
        type: default
    use_position_embeddings: false
    vocab_size: 128256
```

Instead of `type: default` in the `model.base_model.transformer.rotary`, I expected to see `llama3`.

Please fix @RaymondLi0 @jlamypoirier

**Comment by RaymondLi0** [comment]
Hmm this used to work, let me have a look

**Comment by RaymondLi0** [comment]
@jlamypoirier  The logic here https://github.com/ServiceNow/Fast-LLM/blob/b385fb08847962789184b0b320878b12a8ef9bd2/fast_llm/engine/checkpoint/external.py#L220 does not work because `converter.export_name` can be a `tuple[str, ...]`.
That's why I had moved it to `get_nested_dict_value` in https://github.com/ServiceNow/Fast-LLM/pull/55/commits/fc9ef1388fd79b1f74daa5b2ad30deec9c122685
The result is that currently, `value` is set to None for when importing a nested dict value like `rope_scaling.rope_type`

Was there a specific reason to remove that logic from `get_nested_dict_value` ?

**Comment by RaymondLi0** [comment]
I fixed the config comparison tool that should have caught this error in the tests. But another error appeared in the tests, having a look

**Comment by RaymondLi0** [comment]
There remains 2 points:
- an error at https://github.com/ServiceNow/Fast-LLM/blob/229050d0a91a0e3c3ee33f3c1a9fc1e4930ffc6c/tests/test_checkpoint.py#L253 because `model._base_model_config` contains (validated?) default values while `config` doesn't.
- address the following:

> @jlamypoirier The logic here
> 
> https://github.com/ServiceNow/Fast-LLM/blob/b385fb08847962789184b0b320878b12a8ef9bd2/fast_llm/engine/checkpoint/external.py#L220
> 
> does not work because `converter.export_name` can be a `tuple[str, ...]`.
> That's why I had moved it to `get_nested_dict_value` in [fc9ef13](https://github.com/ServiceNow/Fast-LLM/commit/fc9ef1388fd79b1f74daa5b2ad30deec9c122685)
> The result is that currently, `value` is set to None for when importing a nested dict value like `rope_scaling.rope_type`
> Was there a specific reason to remove that logic from `get_nested_dict_value` ?

**Comment by tscholak** [comment]
Thanks @RaymondLi0
Additionally, exporting the checkpoint failed:

```
2024-12-06 17:19:24,749 [Rank 00] Saving export at iteration 1000
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
Unrecognized keys in `rope_scaling` for 'rope_type'='default': {'original_max_position_embeddings', 'factor', 'high_freq_factor', 'low_freq_factor'}
2024-12-06 17:19:24,773 [Rank 00] Saving tensors to /app/fast-llm-tutorial/experiment/export/llama/1000/model_0.safetensors
2024-12-06 17:20:00,939 [Rank 00] Saving tensors to /app/fast-llm-tutorial/experiment/export/llama/1000/model_1.safetensors
[E1206 17:20:24.336092093 ProcessGroupNCCL.cpp:588] [Rank 7] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=308230, OpType=ALLREDUCE, NumelIn=1, NumelOut=1, Timeout(ms)=60000) ran for 60004 milliseconds before timing out.
[E1206 17:20:24.336503904 ProcessGroupNCCL.cpp:1683] [PG  Rank 7] Exception (either an error or timeout) detected by watchdog at work: 308230, last enqueued NCCL work: 308230, last completed NCCL work: 308229.
[E1206 17:20:24.338932817 ProcessGroupNCCL.cpp:588] [Rank 3] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=308230, OpType=ALLREDUCE, NumelIn=1, NumelOut=1, Timeout(ms)=60000) ran for 60008 milliseconds before timing out.
[E1206 17:20:24.339213368 ProcessGroupNCCL.cpp:1683] [PG  Rank 3] Exception (either an error or timeout) detected by watchdog at work: 308230, last enqueued NCCL work: 308230, last completed NCCL work: 308229.
```

File: tests/test_checkpoint.py [hunk 4]
```diff
@@ -241,7 +241,7 @@ def _compare_configs(config_ref, config_test):
 @pytest.mark.depends(on=["test_converted_distributed"])
 def test_load_pretrained_distributed_checkpoint():
     config = TEST_ARCHITECTURE_CONFIG_CLS.from_dict(
-        yaml.safe_load((_CKPT_PATH / ".." / ".." / "config.yaml").open("r")), strict=False
+        yaml.safe_load((_CKPT_PATH / ".." / ".." / "config.yaml").open("r"))["model"]["base_model"], strict=False
```
Comment (jlamypoirier) [inline tests/test_checkpoint.py:244]: I don't understand this one. Wasn't it working before? I'd have caught it in tests...

Comment (RaymondLi0) [reply tests/test_checkpoint.py:244]: The config comparison was broken I think (fixed in https://github.com/ServiceNow/Fast-LLM/pull/55/commits/fa6440e1cf2fc9d0aa4d050e7a17bdcc6e804619), which is why this one wasn't caught

**Comment by jlamypoirier** [comment]
Bugs should be fixed, I but I still haven't checked beyond the tests. One of the tests fail (test_load_distributed_checkpoint_dp2) but it's very minor and unrelated to this PR, looks like something goes wrong with the default init_method_std when loading models.

**Comment by tscholak** [comment]
Thanks @RaymondLi0 and @jlamypoirier for making progress on this. Loading the Llama 3.1 model with the correct rope scaling type works fine now:

```yaml
model:
  base_model:
    transformer:
      rotary:
        type: llama3
        theta: 500000.0
```

However, exporting the model to Llama 3.1 format still fails:

```
2024-12-07 04:15:10,553 [Rank 00] Saving checkpoint at iteration 1000
2024-12-07 04:15:31,546 [Rank 00] Saved checkpoint to /app/fast-llm-tutorial/experiment/checkpoint/1000
2024-12-07 04:15:31,548 [Rank 00] Saving export at iteration 1000
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
`rope_scaling`'s original_max_position_embeddings field must be less than max_position_embeddings, got 8192 and max_position_embeddings=2048
2024-12-07 04:15:31,601 [Rank 00] Saving tensors to /app/fast-llm-tutorial/experiment/export/llama/1000/model_0.safetensors
2024-12-07 04:16:07,999 [Rank 00] Saving tensors to /app/fast-llm-tutorial/experiment/export/llama/1000/model_1.safetensors
[E1207 04:16:31.569811685 ProcessGroupNCCL.cpp:588] [Rank 3] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=308230, OpType=ALLREDUCE, NumelIn=1, NumelOut=1, Timeout(ms)=60000) ran for 60003 milliseconds before timing out.
```

**Comment by tscholak** [comment]
Hi @RaymondLi0, I ran the Big QuickStart guide on LambdaLabs, so:

```yaml
kubectl apply -f - <<EOF
apiVersion: "kubeflow.org/v1"
kind: "PyTorchJob"
metadata:
  name: "fast-llm-train"
spec:
  nprocPerNode: "8"
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      restartPolicy: Never
      template:
        spec:
          tolerations:
            - key: nvidia.com/gpu
              value: "true"
              operator: Equal
              effect: NoSchedule
          containers:
            - name: pytorch
              image: ghcr.io/servicenow/fast-llm:sha-41fa4bd
              resources:
                limits:
                  nvidia.com/gpu: 8
                  rdma/rdma_shared_device_a: 1
                  memory: "1024Gi"
                  cpu:
                requests:
                  nvidia.com/gpu: 8
                  rdma/rdma_shared_device_a: 1
                  memory: "1024Gi"
                  cpu: 128
              command:
                - /bin/bash
                - -c
                - |
                  torchrun --rdzv_backend=static \
                           --rdzv_endpoint=\${MASTER_ADDR}:\${MASTER_PORT} \
                           --node_rank=\${RANK} \
                           --nproc_per_node=\${PET_NPROC_PER_NODE} \
                           --nnodes=\${PET_NNODES} \
                           --max_restarts=0 \
                           --rdzv_conf=timeout=3600 \
                           --no_python \
                           fast-llm train gpt \
                           --config fast-llm-tutorial/train-config.yaml
              env:
                - name: PYTHONHASHSEED
                  value: "0"
                - name: WANDB_API_KEY_PATH
                  value: "/app/fast-llm-tutorial/.wandb_api_key"
                - name: TORCH_NCCL_ASYNC_ERROR_HANDLING
                  value: "1"
                - name: NCCL_DEBUG
                  value: "INFO"
              securityContext:
                capabilities:
                  add:
                    - IPC_LOCK
              volumeMounts:
                - mountPath: /app/fast-llm-tutorial
                  name: fast-llm-inputs
                - mountPath: /dev/shm
                  name: dshm
          volumes:
            - name: fast-llm-inputs
              persistentVolumeClaim:
                claimName: pvc-fast-llm-tutorial
            - name: dshm
              emptyDir:
                medium: Memory
                sizeLimit: "1024Gi"
    Worker:
      replicas: 3
      restartPolicy: Never
      template:
        spec:
          tolerations:
            - key: nvidia.com/gpu
              value: "true"
              operator: Equal
              effect: NoSchedule
          containers:
            - name: pytorch
              image: ghcr.io/servicenow/fast-llm:sha-41fa4bd
              resources:
                limits:
                  nvidia.com/gpu: 8
                  rdma/rdma_shared_device_a: 1
                  memory: "1024Gi"
                  cpu:
                requests:
                  nvidia.com/gpu: 8
                  rdma/rdma_shared_device_a: 1
                  memory: "1024Gi"
                  cpu: 128
              command:
                - /bin/bash
                - -c
                - |
                  torchrun --rdzv_backend=static \
                           --rdzv_endpoint=\${MASTER_ADDR}:\${MASTER_PORT} \
                           --node_rank=\${RANK} \
                           --nproc_per_node=\${PET_NPROC_PER_NODE} \
                           --nnodes=\${PET_NNODES} \
                           --max_restarts=0 \
                           --rdzv_conf=timeout=3600 \
                           --no_python \
                           fast-llm train gpt \
                           --config fast-llm-tutorial/train-config.yaml
              env:
                - name: PYTHONHASHSEED
                  value: "0"
                - name: WANDB_API_KEY_PATH
                  value: "/app/fast-llm-tutorial/.wandb_api_key"
                - name: TORCH_NCCL_ASYNC_ERROR_HANDLING
                  value: "1"
                - name: NCCL_DEBUG
                  value: "INFO"
              securityContext:
                capabilities:
                  add:
                    - IPC_LOCK
              volumeMounts:
                - mountPath: /app/fast-llm-tutorial
                  name: fast-llm-inputs
                - mountPath: /dev/shm
                  name: dshm
          volumes:
            - name: fast-llm-inputs
              persistentVolumeClaim:
                claimName: pvc-fast-llm-tutorial
            - name: dshm
              emptyDir:
                medium: Memory
                sizeLimit: "1024Gi"
EOF
```

with

```yaml
training:
  train_iters: 100_000
  logs:
    interval: 10
  validation:
    iterations: 25
    interval: 1000
  checkpoint:
    interval: 1000
    keep: 5
  test_iters: 0
  export:
    format: llama
    interval: 1_000
  wandb:
    project_name: fast-llm-tutorial
    group_name: Big
    entity_name: null
batch:
  micro_batch_size: 2
  sequence_length: 4096
  batch_size: 512
data:
  format: file
  path: fast-llm-tutorial/dataset/fast_llm_dataset.json
  split: [99, 1, 0]
optimizer:
  weight_decay: 0.1
  beta_1: 0.9
  beta_2: 0.95
  learning_rate:
    base: 6.0e-04
    minimum: 6.0e-05
    decay_style: cosine
    decay_iterations: 100_000
    warmup_iterations: 2000
pretrained:
  format: llama
  path: fast-llm-tutorial/pretrained-model
  model_weights: yes
model:
  base_model:
    transformer:
      use_flash_attention: yes
    cross_entropy_impl: fused
  multi_stage:
    zero_stage: 2
  distributed:
    training_dtype: bf16
run:
  experiment_dir: fast-llm-tutorial/experiment
```

**Review by tscholak — APPROVED** [review]
Confirmed that checkpoint loading and saving works as designed for Llama-3.1-8B.
